egret/model_library/transmission/tx_calc.py

```diff
@@ -547,6 +565,89 @@ def _calculate_pfl_constant(branches,buses,index_set_branch,base_point=BasePoint
+                                 mapping_branch_to_idx=None,
+                                 interfaces=None,
+                                 index_set_interface=None):
+
+    if interfaces is None:
+        assert index_set_interface is None
+    if index_set_interface is None:
+        assert interfaces is None
+
+    _len_bus = len(index_set_bus)
+
+    if mapping_bus_to_idx is None:
+        mapping_bus_to_idx = {bus_n: i for i, bus_n in enumerate(index_set_bus)}
+
+    _len_branch = len(index_set_branch)
+
+    _ref_bus_idx = mapping_bus_to_idx[reference_bus]
+
+    ## check if the network is connected
+    connected = check_network_connection(branches, index_set_branch, index_set_bus, mapping_bus_to_idx)
+
+    if not connected:
+        raise RuntimeError("Network is not connected, cannot use PTDF formulation")
+
+    #(B_d A)
+    J = _calculate_J11(branches,buses,index_set_branch,index_set_bus,mapping_bus_to_idx,base_point,approximation_type=ApproximationType.PTDF)
+    #(A^T)
+    At = calculate_adjacency_matrix_transpose(branches,index_set_branch,index_set_bus,mapping_bus_to_idx)
+
+    ref_bus_mask = np.ones(_len_bus, dtype=bool)
+    ref_bus_mask[_ref_bus_idx] = False
+
+    # M is now (A^T B_d A) with
+    # row and column of reference
+    # bus removed
+    M = (At@J)[ref_bus_mask,:][:,ref_bus_mask]
+
+    # (B_d A) with reference bus column removed
+    B_dA = J[:,ref_bus_mask]
+
+    ## LU factorization
+    MLU_MP = scipy.sparse.linalg.splu(M)
+
+    if interfaces is not None:
+        if mapping_bus_to_idx is None:
+            mapping_branch_to_idx = {branch_n: i for i, branch_n in enumerate(index_set_branch)}
+        I = _calculate_interface_matrix(interfaces, index_set_interface, mapping_branch_to_idx)
+        B_dA_I = I@B_dA
+
+    return MLU_MP, B_dA, ref_bus_mask, B_dA_I, I
+
+def _calculate_interface_matrix(interfaces, index_set_interface, mapping_branch_to_idx):
+    """
+    calculate an interface matrix, where a the rows correspond to each interface, and
+    the columns correspond to each branch
+    """
+    _len_interface = len(index_set_interface)
+    _len_branch = len(mapping_branch_to_idx)
+
+    row = []
+    col = []
+    data = []
+
+    for idx, i_n in enumerate(index_set_interface):
+        interface = interfaces[i_n]
+        for l, val in zip(interface['lines'], interface['line_orientation']):
+            if val == 0:
+                continue
+            elif val == -1 or val == 1:
+                branch_idx = mapping_branch_to_idx[l]
+                row.append(idx)
+                col.append(branch_idx)
+                data.append(val)
+            else:
+                ## TODO: do we need enforce this requirement?
+                raise Exception("Interface {} has line {} with line_orientation {} "
+                        "not in [-1,0,1].".format(i_n, l, val))
+    I = sp.coo_matrix((data,(row,col)),shape=(_len_interface,_len_branch))
+    return I.tocsr()
```

Comment:
Is there a need to convert this to csr? In my experience, coo matrix dot products are just as efficient, if not more so.

Reply:
No, I don't think so. We don't do much with this matrix, so I'll leave it as a `coo_matrix`.